### PR TITLE
Fix Xaml compilation generating an empty InitializeComponent method when App contains a single item

### PR DESCRIPTION
**Pull request by ThomasGoulet73**
Fixes dotnet/wpf#2543
Fixes dotnet/wpf#2926
Fixes dotnet/wpf#4457
Fixes dotnet/wpf#4538
Fixes dotnet/wpf#6198

## Description
There were many bug reports of different type that were all caused by the same bug. Xaml compilation generating an empty InitializeComponent method when App contains a single item.

Some context:
When generating App.InitializeComponent for a Xaml where there is only the StartupUri property, it will skip loading Baml in App.InitializeComponent and only set the StartupUri property in the generated code for performance reasons. There is currently a bug when detecting if the StartupUri property is the only property that would be loaded from the Baml.

See https://github.com/dotnet/wpf/pull/8853#discussion_r1507014370 for an explanation of the changes in this PR.

## Customer Impact
Should fix issues linked at the top (Might be more since the bug does not always surface the same way)

## Regression
Yes and no, the bug was not present in non-SDK style .Net Framework projects.

## Testing
Manual testing using the samples provided in the issues and manual testing with my own samples with multiple Xaml combination.

## Risk
Low to medium. Worst case scenario this PR could cause existing project that were not loading Baml to now load Baml which could cause slight perf regression but I wasn't able to cause this in my testing.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/wpf/pull/8853)

### CODE REVIEW

File: src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs [hunk 1]
```diff
@@ -517,7 +517,7 @@ public override void WriteElementEnd(XamlElementEndNode xamlEndObjectNode)
                         xmlReader.MoveToAttribute(attrName);
                 }
             }
-            else if (!_compiler.IsBamlNeeded && !_compiler.ProcessingRootContext && _compiler.IsCompilingEntryPointClass && xmlReader.Depth > 0)
```
Comment (ThomasGoulet73) [inline src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: This code is called too early to call `_compiler.ProcessingRootContext`.

Basically, there are two steps involved here:
1. Reading a Xaml element.
2. Parsing the Xaml element.

`_compiler.ProcessingRootContext` will only be set in step 2 but this code here is ran in step 1. This means that we get its value before it was set. Since both steps are ran for each Xaml element, when we have 2 or more element we execute this code here with `_compiler.ProcessingRootContext` initialized.

We can remove the check to `_compiler.ProcessingRootContext` because it was already superfluous because the check here `ProcessedRootElement`:
https://github.com/dotnet/wpf/blob/c9a4c6b4a3974df86ffac8d21af1b488c63c630a/src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs#L414
basically does the same check as `_compiler.ProcessingRootContext` but `ProcessedRootElement` is set in step 1 instead of step 2.

I hope my explanation was clear enough to explain my reasoning. Feel free to ask question and I'll try to answer though the Xaml compilation process is very complex and I might not be able to properly explain it.  It makes sense in my head 😄.

Comment (anjali-wpf) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: Thanks for your contribution @ThomasGoulet73 , this change simplifies the code by removing a redundant check to _compiler.ProcessingRootContext, which is called too early in the XAML compilation process to have its value set. The existing check to ProcessedRootElement already serves the same purpose and is set earlier in the process, making the _compiler.ProcessingRootContext check unnecessary. This improves code clarity and avoids potential confusion.
Possibly you can add this in the description as well for better understanding.

Comment (ThomasGoulet73) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: @anjali-wpf I found it easier to comment on the changes on this specific line. I added a link to this comment in the description to improve clarity.

**Review by anjali-wpf — APPROVED** [review]

**Review by hiteshwpfmsft — APPROVED** [review]

File: src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs (hunk 1 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: @ThomasGoulet73, can you give me a general idea of where this function is being called in both the stages ?

Comment (ThomasGoulet73) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: Sure, I just need to refresh my memory a bit and I'll get back to you.

Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: @ThomasGoulet73, I understood what you meant. 
From my understanding, when TokenReder.Read method is called in XamlReaderHelper, we get to this point where the ProcessingRootContext is not yet set, and when we do _compiler.StartElement call from WriteElementStart in ParserExtension, then this variable get's set. That's why when we have two nodes, it is processed correctly.

LGTM.

**Review by dipeshmsft — APPROVED** [review]

File: src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs (hunk 1 above)
Comment (ThomasGoulet73) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/MarkupCompiler/ParserExtension.cs:520]: Thanks for merging. Sorry I didn't have time to reply. I started looking into it but hadn't finished yet but from what I saw my understanding matched what you said.
